File: scala-kata/src/main/scala/org/suggs/katas/bankaccount/Account.scala
package org.suggs.katas.bankaccount

import org.suggs.katas.bankaccount.Money.anAmountOf

object Account {
  def anAccountWith(money: Money) = Account(money)

  def anEmptyAccount() = Account(anAmountOf(0.0d))

}

case class Account(amount: Money) {
  def withdraw(money: Money) = Account(amount.less(money))

  def deposit(money: Money) = Account(amount.add(money))

}

File: java-kata/src/main/java/org/suggs/katas/bankaccount/Account.java
package org.suggs.katas.bankaccount;

import lombok.EqualsAndHashCode;

import static org.suggs.katas.bankaccount.Money.anAmountOf;

@EqualsAndHashCode
public class Account {
    private Money balance;

    public static Account anAccountWith(Money amount) {
        return new Account(amount);
    }

    public static Account anEmptyAccount() {
        return new Account(anAmountOf(0.0d));
    }

    private Account(Money anAmount) {
        this.balance = anAmount;
    }

    public void deposit(final Money anAmount) {
        balance.add(anAmount);
    }

    public void withdraw(final Money anAmount) {
        balance.less(anAmount);
    }

    public void transfer(Money money, Account destinationAccount) {
        destinationAccount.deposit(money);
        this.withdraw(money);
    }
}
